src/compiler/checker.ts

```diff
@@ -8987,6 +8978,45 @@ module ts {
             }
         }
 
+        function checkElementTypeOfArrayOrString(arrayOrStringType: Type, expressionForError: Expression): Type {
+            Debug.assert(languageVersion < ScriptTarget.ES6);
+            var isJustString = allConstituentTypesHaveKind(arrayOrStringType, TypeFlags.StringLike);
+            
+            // Check isJustString because removeTypesFromUnionType will only remove types if it doesn't result
+            // in an emptyObjectType. In this case, we actually do want the emptyObjectType.
+            var arrayType = isJustString ? emptyObjectType : removeTypesFromUnionType(arrayOrStringType, TypeFlags.StringLike, /*isTypeOfKind*/ true);
+            var hasStringConstituent = arrayOrStringType !== emptyObjectType && arrayOrStringType !== arrayType;
+
+            var reportedError = false;
+            if (hasStringConstituent && languageVersion < ScriptTarget.ES5) {
+                error(expressionForError, Diagnostics.Using_a_string_in_a_for_of_statement_is_only_supported_in_ECMAScript_5_and_higher);
+                reportedError = true;
+            }
+
+            if (isJustString) {
+                return stringType;
+            }
+
+            if (!isArrayLikeType(arrayType)) {
+                if (!reportedError) {
+                    error(expressionForError, Diagnostics.Type_0_is_not_an_array_type, typeToString(arrayType));
+                }
+                return hasStringConstituent ? stringType : unknownType;
+            }
+
+            var arrayElementType = getIndexTypeOfType(arrayType, IndexKind.Number) || unknownType;
+            if (hasStringConstituent) {
+                // This is just an optimization for the case where arrayOrStringType is string | string[]

```

Comment:
Could you add, at the top of this function, the types of types you want this fnction to be able to handle, and what final type will be returned?

Note: this function never seems to return 'undefined'.  But i thought you used || earlier to default to hte 'any' type of you got 'undefined' back from this.